File: src/generativeUI/world-monitor-main/mcp_server/graph/kuzu_store.py
# SPDX-License-Identifier: AGPL-3.0-only
# SPDX-FileCopyrightText: 2024 SAP SE
"""
KùzuDB Graph Store for World Monitor MCP Server.

Maintains a property graph of geopolitical monitoring entities used to
enrich alert and metrics responses with correlated-event and service context.

Schema
------
Node tables
  GeoEvent     – a geopolitical/natural event (conflict, disaster, cyber, …)
  ServiceNode  – a registered MCP service or monitored endpoint
  AlertRecord  – a named alert with severity and originating service

Relationship tables
  TRIGGERS_ALERT  – GeoEvent → AlertRecord
  MONITORED_BY    – GeoEvent → ServiceNode  (which service tracks this event)
  AFFECTS_SERVICE – AlertRecord → ServiceNode

Usage
-----
    from mcp_server.graph.kuzu_store import get_kuzu_store

    store = get_kuzu_store()
    store.ensure_schema()
    store.upsert_event("evt-001", "conflict", "Ukraine", "Eastern Europe")
    store.upsert_service("world-monitor-mcp", "http://localhost:9170/mcp", "monitoring")
    store.upsert_alert("alt-001", "High Conflict Activity", "critical", "world-monitor-mcp")
    store.link_event_alert("evt-001", "alt-001")
    ctx = store.get_event_context("evt-001")
"""

import os
import sys
from typing import Any

KUZU_DB_PATH: str = os.environ.get("KUZU_DB_PATH", ":memory:").strip()

_singleton: "KuzuStore | None" = None


def get_kuzu_store() -> "KuzuStore":
    global _singleton
    if _singleton is None:
        _singleton = KuzuStore(KUZU_DB_PATH)
    return _singleton


def _reset_kuzu_store() -> None:
    """Reset the singleton (test use only)."""
    global _singleton
    _singleton = None


class KuzuStore:
    """Embedded KùzuDB graph store for world-monitor graph-RAG features."""

    def __init__(self, db_path: str = ":memory:") -> None:
        self.db_path = db_path
        self._db: Any = None
        self._conn: Any = None
        self._available = self._init_db()
        self._schema_ready = False

    # ------------------------------------------------------------------
    # M1 – Initialisation
    # ------------------------------------------------------------------

    def _init_db(self) -> bool:
        try:
            import kuzu  # type: ignore[import]
            self._db = kuzu.Database(self.db_path)
            self._conn = kuzu.Connection(self._db)
            print(f"INFO: KùzuDB initialised at '{self.db_path}'", file=sys.stderr)
            return True
        except Exception as exc:
            msg = str(exc)
            if "No module named" in msg or "ModuleNotFoundError" in msg:
                print(
                    "WARNING: kuzu Python package not installed; graph-RAG features disabled. "
                    "Add 'kuzu>=0.7.0' to mcp_server/requirements.txt to enable.",
                    file=sys.stderr,
                )
            else:
                print(f"WARNING: KùzuDB init failed: {msg}", file=sys.stderr)
            return False

    def available(self) -> bool:
        return self._available

    def ensure_schema(self) -> None:
        if not self._available or self._schema_ready:
            return
        stmts = [
            # Node: a geopolitical / natural / cyber event
            """CREATE NODE TABLE IF NOT EXISTS GeoEvent (
                event_id    STRING,
                event_type  STRING,
                country     STRING,
                region      STRING,
                PRIMARY KEY (event_id)
            )""",
            # Node: a registered MCP service or monitored endpoint
            """CREATE NODE TABLE IF NOT EXISTS ServiceNode (
                name        STRING,
                endpoint    STRING,
                category    STRING,
                PRIMARY KEY (name)
            )""",
            # Node: an alert record
            """CREATE NODE TABLE IF NOT EXISTS AlertRecord (
                alert_id    STRING,
                name        STRING,
                severity    STRING,
                service     STRING,
                PRIMARY KEY (alert_id)
            )""",
            # Relationship: event triggers an alert
            """CREATE REL TABLE IF NOT EXISTS TRIGGERS_ALERT (
                FROM GeoEvent TO AlertRecord
            )""",
            # Relationship: event monitored by a service
            """CREATE REL TABLE IF NOT EXISTS MONITORED_BY (
                FROM GeoEvent TO ServiceNode
            )""",
            # Relationship: alert affects a service
            """CREATE REL TABLE IF NOT EXISTS AFFECTS_SERVICE (
                FROM AlertRecord TO ServiceNode
            )""",
        ]
        for stmt in stmts:
            try:
                self._exec(stmt)
            except Exception as exc:
                msg = str(exc)
                if "already exists" not in msg:
                    print(f"KùzuDB schema stmt skipped ({msg[:80]})", file=sys.stderr)
        self._schema_ready = True

    # ------------------------------------------------------------------
    # Low-level query helper
    # ------------------------------------------------------------------

    def _exec(self, cypher: str, params: dict | None = None) -> Any:
        if params:
            return self._conn.execute(cypher, params)
        return self._conn.execute(cypher)

    # ------------------------------------------------------------------
    # M2 helpers – upsert operations called by kuzu_index tool
    # ------------------------------------------------------------------

    def upsert_event(
        self,
        event_id: str,
        event_type: str = "",
        country: str = "",
        region: str = "",
    ) -> None:
        if not self._available:
            return
        try:
            self._exec(
                "MERGE (e:GeoEvent {event_id: $id}) "
                "SET e.event_type = $etype, e.country = $country, e.region = $region",
                {"id": event_id, "etype": event_type, "country": country, "region": region},
            )
        except Exception:
            try:
                self._exec(
                    "CREATE (e:GeoEvent {event_id: $id, event_type: $etype, "
                    "country: $country, region: $region})",
                    {"id": event_id, "etype": event_type, "country": country, "region": region},
                )
            except Exception as exc2:
                print(f"upsert_event failed for {event_id}: {exc2}", file=sys.stderr)

    def upsert_service(self, name: str, endpoint: str = "", category: str = "") -> None:
        if not self._available:
            return
        try:
            self._exec(
                "MERGE (s:ServiceNode {name: $name}) "
                "SET s.endpoint = $ep, s.category = $cat",
                {"name": name, "ep": endpoint, "cat": category},
            )
        except Exception:
            try:
                self._exec(
                    "CREATE (s:ServiceNode {name: $name, endpoint: $ep, category: $cat})",
                    {"name": name, "ep": endpoint, "cat": category},
                )
            except Exception as exc2:
                print(f"upsert_service failed for {name}: {exc2}", file=sys.stderr)

    def upsert_alert(
        self,
        alert_id: str,
        name: str = "",
        severity: str = "warning",
        service: str = "",
    ) -> None:
        if not self._available:
            return
        try:
            self._exec(
                "MERGE (a:AlertRecord {alert_id: $id}) "
                "SET a.name = $name, a.severity = $sev, a.service = $svc",
                {"id": alert_id, "name": name, "sev": severity, "svc": service},
            )
        except Exception:
            try:
                self._exec(
                    "CREATE (a:AlertRecord {alert_id: $id, name: $name, "
                    "severity: $sev, service: $svc})",
                    {"id": alert_id, "name": name, "sev": severity, "svc": service},
                )
            except Exception as exc2:
                print(f"upsert_alert failed for {alert_id}: {exc2}", file=sys.stderr)

    def link_event_alert(self, event_id: str, alert_id: str) -> None:
        if not self._available:
            return
        try:
            self._exec(
                "MATCH (e:GeoEvent {event_id: $eid}), (a:AlertRecord {alert_id: $aid}) "
                "CREATE (e)-[:TRIGGERS_ALERT]->(a)",
                {"eid": event_id, "aid": alert_id},
            )
        except Exception as exc:
            print(f"link_event_alert failed {event_id}->{alert_id}: {exc}", file=sys.stderr)

    def link_event_service(self, event_id: str, service_name: str) -> None:
        if not self._available:
            return
        try:
            self._exec(
                "MATCH (e:GeoEvent {event_id: $eid}), (s:ServiceNode {name: $sname}) "
                "CREATE (e)-[:MONITORED_BY]->(s)",
                {"eid": event_id, "sname": service_name},
            )
        except Exception as exc:
            print(f"link_event_service failed {event_id}->{service_name}: {exc}", file=sys.stderr)

    def link_alert_service(self, alert_id: str, service_name: str) -> None:
        if not self._available:
            return
        try:
            self._exec(
                "MATCH (a:AlertRecord {alert_id: $aid}), (s:ServiceNode {name: $sname}) "
                "CREATE (a)-[:AFFECTS_SERVICE]->(s)",
                {"aid": alert_id, "sname": service_name},
            )
        except Exception as exc:
            print(f"link_alert_service failed {alert_id}->{service_name}: {exc}", file=sys.stderr)

    # ------------------------------------------------------------------
    # M3 helper – raw Cypher query used by kuzu_query tool
    # ------------------------------------------------------------------

    def run_query(self, cypher: str, params: dict | None = None) -> list[dict]:
        if not self._available:
            return []
        try:
            result = self._exec(cypher, params or {})
            col_names: list[str] = result.get_column_names() if hasattr(result, "get_column_names") else []
            rows: list[dict] = []
            while result.has_next():
                row = result.get_next()
                rows.append(dict(zip(col_names, row)))
            return rows
        except Exception as exc:
            print(f"KùzuDB query failed: {exc}", file=sys.stderr)
            return []

    # ------------------------------------------------------------------
    # M4 helper – graph context enrichment for get_alerts
    # ------------------------------------------------------------------

    def get_event_context(self, event_id: str) -> list[dict]:
        """Return related alerts and services for a given event_id."""
        if not self._available:
            return []

        alert_rows = self.run_query(
            "MATCH (e:GeoEvent {event_id: $id})-[:TRIGGERS_ALERT]->(a:AlertRecord) "
            "RETURN a.alert_id AS alert_id, a.name AS name, a.severity AS severity, "
            "'triggered_alert' AS relation LIMIT 10",
            {"id": event_id},
        )
        service_rows = self.run_query(
            "MATCH (e:GeoEvent {event_id: $id})-[:MONITORED_BY]->(s:ServiceNode) "
            "RETURN s.name AS service_name, s.endpoint AS endpoint, "
            "'monitored_by' AS relation LIMIT 10",
            {"id": event_id},
        )
        return alert_rows + service_rows

    def get_alert_context(self, alert_id: str) -> list[dict]:
        """Return correlated events and affected services for a given alert_id."""
        if not self._available:
            return []

        event_rows = self.run_query(
            "MATCH (e:GeoEvent)-[:TRIGGERS_ALERT]->(a:AlertRecord {alert_id: $id}) "
            "RETURN e.event_id AS event_id, e.event_type AS event_type, "
            "e.country AS country, 'correlated_event' AS relation LIMIT 10",
            {"id": alert_id},
        )
        service_rows = self.run_query(
            "MATCH (a:AlertRecord {alert_id: $id})-[:AFFECTS_SERVICE]->(s:ServiceNode) "
            "RETURN s.name AS service_name, s.endpoint AS endpoint, "
            "'affects_service' AS relation LIMIT 10",
            {"id": alert_id},
        )
        return event_rows + service_rows

File: src/generativeUI/ui5-webcomponents-ngx-main/mcp-server/src/kuzu-store.ts
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2024 SAP SE
/**
 * KùzuDB Graph Store for UI5 Web Components Angular MCP Server.
 *
 * Maintains a property graph of UI5 component relationships used to enrich
 * template-generation responses with related-component and co-usage context.
 *
 * Schema
 * ------
 * Node tables
 *   Ui5Component   – a UI5 Web Component (tag name is the primary key)
 *   AngularModule  – an Angular NgModule that exposes components
 *   ComponentSlot  – a named slot exposed by a component
 *
 * Relationship tables
 *   BELONGS_TO     – Ui5Component → AngularModule
 *   HAS_SLOT       – Ui5Component → ComponentSlot
 *   CO_USED_WITH   – Ui5Component → Ui5Component (template co-usage)
 *
 * Usage
 * -----
 *   import { getKuzuStore } from './kuzu-store';
 *
 *   const store = getKuzuStore();
 *   await store.ensureSchema();
 *   await store.upsertComponent('ui5-button', 'Ui5ButtonModule', ['default', 'icon']);
 *   await store.linkCoUsage('ui5-button', 'ui5-dialog');
 *   const ctx = await store.getComponentContext('ui5-button');
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

const KUZU_DB_PATH = (process.env['KUZU_DB_PATH'] ?? ':memory:').trim();

let _singleton: KuzuStore | null = null;

export function getKuzuStore(): KuzuStore {
  if (!_singleton) {
    _singleton = new KuzuStore(KUZU_DB_PATH);
  }
  return _singleton;
}

/** Reset the singleton (test use only). */
export function _resetKuzuStore(): void {
  _singleton = null;
}

export interface GraphContextEntry {
  relation: string;
  [key: string]: unknown;
}

export class KuzuStore {
  private dbPath: string;
  private db: any = null;
  private conn: any = null;
  private _available = false;
  private _schemaReady = false;

  constructor(dbPath: string = ':memory:') {
    this.dbPath = dbPath;
    this._available = this.initDb();
  }

  // ---------------------------------------------------------------------------
  // M1 – Initialisation
  // ---------------------------------------------------------------------------

  private initDb(): boolean {
    try {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const kuzu = require('kuzu') as { Database: new (p: string) => any; Connection: new (db: any) => any };
      this.db = new kuzu.Database(this.dbPath);
      this.conn = new kuzu.Connection(this.db);
      console.info(`INFO: KùzuDB initialised at '${this.dbPath}'`);
      return true;
    } catch (e: unknown) {
      const msg = e instanceof Error ? e.message : String(e);
      if (msg.includes('Cannot find module')) {
        console.warn(
          'WARNING: kuzu npm package not installed; graph-RAG features disabled. ' +
          "Add 'kuzu' to mcp-server/package.json dependencies to enable.",
        );
      } else {
        console.warn(`WARNING: KùzuDB init failed: ${msg}`);
      }
      return false;
    }
  }

  available(): boolean {
    return this._available;
  }

  async ensureSchema(): Promise<void> {
    if (!this._available || this._schemaReady) return;
    const stmts = [
      // Node: a UI5 web component
      `CREATE NODE TABLE IF NOT EXISTS Ui5Component (
        tag_name      STRING,
        angular_module STRING,
        npm_module    STRING,
        PRIMARY KEY (tag_name)
      )`,
      // Node: an Angular NgModule
      `CREATE NODE TABLE IF NOT EXISTS AngularModule (
        module_name STRING,
        package     STRING,
        PRIMARY KEY (module_name)
      )`,
      // Node: a named slot of a component
      `CREATE NODE TABLE IF NOT EXISTS ComponentSlot (
        slot_id   STRING,
        tag_name  STRING,
        slot_name STRING,
        PRIMARY KEY (slot_id)
      )`,
      // Relationship: component belongs to Angular module
      `CREATE REL TABLE IF NOT EXISTS BELONGS_TO (
        FROM Ui5Component TO AngularModule
      )`,
      // Relationship: component exposes a slot
      `CREATE REL TABLE IF NOT EXISTS HAS_SLOT (
        FROM Ui5Component TO ComponentSlot
      )`,
      // Relationship: two components frequently appear together in templates
      `CREATE REL TABLE IF NOT EXISTS CO_USED_WITH (
        FROM Ui5Component TO Ui5Component,
        weight INTEGER
      )`,
    ];
    for (const stmt of stmts) {
      try {
        await this.exec(stmt);
      } catch (e: unknown) {
        // IF NOT EXISTS makes this idempotent; log debug only
        const msg = e instanceof Error ? e.message : String(e);
        if (!msg.includes('already exists')) {
          console.debug(`KùzuDB schema stmt skipped (${msg.slice(0, 80)})`);
        }
      }
    }
    this._schemaReady = true;
  }

  // ---------------------------------------------------------------------------
  // Low-level query helper
  // ---------------------------------------------------------------------------

  private async exec(cypher: string, params: Record<string, unknown> = {}): Promise<any> {
    return new Promise((resolve, reject) => {
      try {
        const result = this.conn.execute(cypher, params);
        resolve(result);
      } catch (e) {
        reject(e);
      }
    });
  }

  // ---------------------------------------------------------------------------
  // M2 helpers – upsert operations called by kuzu_index tool
  // ---------------------------------------------------------------------------

  async upsertComponent(
    tagName: string,
    angularModule: string,
    npmModule = '',
    slots: string[] = [],
  ): Promise<void> {
    if (!this._available) return;
    // Upsert Ui5Component node
    try {
      await this.exec(
        'MERGE (c:Ui5Component {tag_name: $tag}) ' +
        'SET c.angular_module = $mod, c.npm_module = $npm',
        { tag: tagName, mod: angularModule, npm: npmModule },
      );
    } catch {
      try {
        await this.exec(
          'CREATE (c:Ui5Component {tag_name: $tag, angular_module: $mod, npm_module: $npm})',
          { tag: tagName, mod: angularModule, npm: npmModule },
        );
      } catch (e2) {
        console.debug(`upsertComponent node failed for ${tagName}: ${String(e2)}`);
      }
    }
    // Upsert AngularModule node
    try {
      await this.exec(
        'MERGE (m:AngularModule {module_name: $mod})',
        { mod: angularModule },
      );
    } catch {
      try {
        await this.exec(
          'CREATE (m:AngularModule {module_name: $mod, package: $pkg})',
          { mod: angularModule, pkg: '@ui5/webcomponents-ngx' },
        );
      } catch (e2) {
        console.debug(`upsertComponent module failed for ${angularModule}: ${String(e2)}`);
      }
    }
    // BELONGS_TO edge
    try {
      await this.exec(
        'MATCH (c:Ui5Component {tag_name: $tag}), (m:AngularModule {module_name: $mod}) ' +
        'CREATE (c)-[:BELONGS_TO]->(m)',
        { tag: tagName, mod: angularModule },
      );
    } catch (e) {
      console.debug(`BELONGS_TO edge failed ${tagName}->${angularModule}: ${String(e)}`);
    }
    // Slot nodes + HAS_SLOT edges
    for (const slotName of slots) {
      const slotId = `${tagName}::${slotName}`;
      try {
        await this.exec(
          'MERGE (s:ComponentSlot {slot_id: $sid}) SET s.tag_name = $tag, s.slot_name = $slot',
          { sid: slotId, tag: tagName, slot: slotName },
        );
      } catch {
        try {
          await this.exec(
            'CREATE (s:ComponentSlot {slot_id: $sid, tag_name: $tag, slot_name: $slot})',
            { sid: slotId, tag: tagName, slot: slotName },
          );
        } catch (e2) {
          console.debug(`slot node failed ${slotId}: ${String(e2)}`);
        }
      }
      try {
        await this.exec(
          'MATCH (c:Ui5Component {tag_name: $tag}), (s:ComponentSlot {slot_id: $sid}) ' +
          'CREATE (c)-[:HAS_SLOT]->(s)',
          { tag: tagName, sid: slotId },
        );
      } catch (e) {
        console.debug(`HAS_SLOT edge failed ${tagName}->${slotId}: ${String(e)}`);
      }
    }
  }

  async linkCoUsage(tagA: string, tagB: string, weight = 1): Promise<void> {
    if (!this._available) return;
    try {
      await this.exec(
        'MATCH (a:Ui5Component {tag_name: $a}), (b:Ui5Component {tag_name: $b}) ' +
        'CREATE (a)-[:CO_USED_WITH {weight: $w}]->(b)',
        { a: tagA, b: tagB, w: weight },
      );
    } catch (e) {
      console.debug(`CO_USED_WITH edge failed ${tagA}->${tagB}: ${String(e)}`);
    }
  }

  // ---------------------------------------------------------------------------
  // M3 helper – raw Cypher query used by kuzu_query tool
  // ---------------------------------------------------------------------------

  async runQuery(cypher: string, params: Record<string, unknown> = {}): Promise<Record<string, unknown>[]> {
    if (!this._available) return [];
    try {
      const result = await this.exec(cypher, params);
      const resultObj = Array.isArray(result) ? result[0] : result;
      if (!resultObj) return [];
      const colNames: string[] = resultObj.getColumnNames ? resultObj.getColumnNames() : [];
      const rows: Record<string, unknown>[] = [];
      while (resultObj.hasNext && resultObj.hasNext()) {
        const row: unknown[] = resultObj.getNext ? resultObj.getNext() : [];
        const record: Record<string, unknown> = {};
        colNames.forEach((col: string, i: number) => { record[col] = row[i]; });
        rows.push(record);
      }
      return rows;
    } catch (e: unknown) {
      console.warn(`KùzuDB query failed: ${String(e)}`);
      return [];
    }
  }

  // ---------------------------------------------------------------------------
  // M4 helper – graph context enrichment for generate_angular_template
  // ---------------------------------------------------------------------------

  async getComponentContext(tagName: string): Promise<GraphContextEntry[]> {
    if (!this._available) return [];

    const moduleRows = await this.runQuery(
      'MATCH (c:Ui5Component {tag_name: $tag})-[:BELONGS_TO]->(m:AngularModule) ' +
      'RETURN m.module_name AS module_name, \'angular_module\' AS relation',
      { tag: tagName },
    );

    const slotRows = await this.runQuery(
      'MATCH (c:Ui5Component {tag_name: $tag})-[:HAS_SLOT]->(s:ComponentSlot) ' +
      'RETURN s.slot_name AS slot_name, \'slot\' AS relation LIMIT 10',
      { tag: tagName },
    );

    const coUsedRows = await this.runQuery(
      'MATCH (a:Ui5Component {tag_name: $tag})-[:CO_USED_WITH]->(b:Ui5Component) ' +
      'RETURN b.tag_name AS co_component, b.angular_module AS co_module, ' +
      '\'co_used_with\' AS relation LIMIT 10',
      { tag: tagName },
    );

    return [
      ...(moduleRows as GraphContextEntry[]),
      ...(slotRows as GraphContextEntry[]),
      ...(coUsedRows as GraphContextEntry[]),
    ];
  }

  async getTemplateContext(tagNames: string[]): Promise<Record<string, GraphContextEntry[]>> {
    const result: Record<string, GraphContextEntry[]> = {};
    for (const tag of tagNames) {
      result[tag] = await this.getComponentContext(tag);
    }
    return result;
  }
}
